### PR TITLE
Fixed the issue #6791, allowing Columns in hstack table operation.

### LINKED ISSUE #6791: hstack requires a sequence of Tables or Rows
I am horizontally stacking a Column and a Table and get this error:
```
.../anaconda2/lib/python2.7/site-packages/astropy/table/operations.pyc in _get_list_of_tables(tables)
     50     # Make sure each thing is a Table or Row
     51     if any(not isinstance(x, (Table, Row)) for x in tables) or len(tables) == 0:
---> 52         raise TypeError('`tables` arg must be a Table or sequence of Tables or Rows')
     53 
     54     # Convert any Rows to Tables

TypeError: `tables` arg must be a Table or sequence of Tables or Rows
```
hstack is incorrectly requiring a Row instead of a Column.

### PR DESCRIPTION
**Pull request by hrishikeshgoyal**
Allowed Columns to appear in list of tables argument to be passed in hstack.
special handling of column object in the list of tables is required.. to avoid that.. I table out of every single column object present in the table list.

Sorry for the making a new pull request as earlier I forgot to sync my repository to current astropy repo(my repository was months old).

EDIT: Fix #6791, supersede #6815

### CODE REVIEW

**Comment by AustereCuriosity** [comment]
@hrishikeshgoyal You don't need to close a PR and open a new one because your master is outdated. You can just rebase your branch by fetching the latest master. This would probably help in the same: http://docs.astropy.org/en/latest/development/workflow/development_workflow.html#how-to-rebase

**Comment by AustereCuriosity** [comment]
I took a quick look at the PR and I think that ``Column`` was purposely left out previously (do wait for other reviews; the following is based on my understanding of what the PR does).

Stacking a list of ``Columns`` horizontally seems as good as ``add_column(s)`` function in ``Table`` (with the exception of adding columns with different number of entries, but I feel that might not be so necessary). Also, ``hstack`` is supposed to be a high level table operation; ``Table`` seems to be a better place for stacking columns.

Another concern is that this function is also used in ``vstack`` and vertical stacking of individual columns seems as good as appending values.

Nevertheless, @taldcroft is the best person to tell you whether this is required or not. Also, there maybe some test cases that I am missing out. If you could think of those and if they do get solved by this PR, please add tests for the same. It becomes easier to review once there are tests 😄

**Comment by hrishikeshgoyal** [comment]
@AustereCuriosity hey.. even after rebase, I saw.. there were some checks failing. and at the moment I wasn't sure  about the cause of failure. So I thought of making a new PR. Anyways thanks :)
..
>>> Stacking a list of Columns horizontally seems as good as add_column(s) function in Table (with the exception of adding columns with different number of entries, but I feel that might not be so necessary).
The usability of including columns in `hstack` becomes non trivial when there are multiple tables and multiple columns. Instead of doing `add_column(s)` to different tables and then horizontally stack the tables, it does the same in a single operation.

File: astropy/table/operations.py [hunk 1]
```diff
@@ -38,18 +38,21 @@ def _get_list_of_tables(tables):
     Check that tables is a Table or sequence of Tables.  Returns the
```
[the remark this answers is not in the document]
Comment (hrishikeshgoyal) [reply astropy/table/operations.py]: sure. I'll do that.

**Comment by AustereCuriosity** [comment]
@hrishikeshgoyal I noticed the issue afterwards and then updated the comment. This will definitely be useful!

**Comment by hrishikeshgoyal** [comment]
@AustereCuriosity  Any documentation or quick guide for writing tests?

File: CHANGES.rst [hunk 2]
```diff
@@ -295,6 +295,8 @@ astropy.stats
 astropy.table
 ^^^^^^^^^^^^^
 
+- Allowed Columns to appear in the parameter 'tables' passed to hstack table operation.[#6791]
```
Comment (AustereCuriosity) [inline CHANGES.rst]: Always use double backquotes in changelog; that is, ``tables``. Also, the number at the end denotes the PR number NOT the issue it solves; use [#6818]. Also, a space after ``.`` (trivial, but to keep with the format of the file).

Probably best to write: 
Allowed ``Column`` to be passed as valid input to the ``hstack`` Table operation. [#6818]

File: astropy/table/operations.py [hunk 3]
```diff
@@ -236,7 +236,7 @@ def vstack(tables, join_type='outer', metadata_conflicts='warn'):
 
     Parameters
     ----------
-    tables : Table or list of Table objects
+    tables : Table or list of Table and/or Row objects
```
Comment (AustereCuriosity) [inline astropy/table/operations.py]: You've made a distinction between ``hstack`` and ``vstack`` as follows:
``Column`` is allowed for ``hstack`` and ``Row`` is allowed for ``vstack`` which makes sense. But the issue is that both call the same function ``_get_list_of_tables`` to validate the input and hence both will accept Rows as well as Columns. If the distinction is necessary, you will have to validate the input accordingly; probably by using a keyword argument in ``_get_list_of_tables`` to make it backwards compatible or separately checking for Column for ``hstack`` (just a suggestion).

File: astropy/table/operations.py [hunk 4]
```diff
@@ -35,21 +35,24 @@ def _merge_table_meta(out, tables, metadata_conflicts='warn'):
...
+    # Make sure each thing is a Table or Row or Column
+    if any(not isinstance(x, (Table, Row, Column)) for x in tables) or len(tables) == 0:
+        raise TypeError('`tables` arg must be a Table or sequence of Tables or Rows or Column')
 
     # Convert any Rows to Tables
-    tables = [(x if isinstance(x, Table) else Table(x)) for x in tables]
+    tables = [(Table(x) if isinstance(x, Row) else x) for x in tables]
+
+    # Convert any Columns to Tables
+    tables = [(Table([x]) if isinstance(x, Column) else x) for x in tables]
```
Comment (taldcroft) [inline astropy/table/operations.py]: In order to allow for mixin columns this needs to be a little more complicated.
```
from .table import Table, Row, has_info_class, BaseColumnClass
```
Then the test is not `isinstance(x, Column)` but `has_info_class(x, BaseColumnClass)`.

File: astropy/table/operations.py (hunk 3 above)
Comment (hrishikeshgoyal) [reply astropy/table/operations.py]: got the point.. will work on it.

File: astropy/table/operations.py (hunk 4 above)
Comment (hrishikeshgoyal) [reply astropy/table/operations.py]: hi.. I am checking this.. and will make the appropriate changes.

**Comment by astrofrog** [comment]
This PR was never finished and required tests. If someone wants to implement this one day in future they can open a new PR with the commits from here and add tests. No need to keep this open since there is https://github.com/astropy/astropy/issues/6791 to keep track of it.

**Comment by bsipocz** [comment]
@astrofrog - actually I was just running into this issue of hstacking columns when prototyping timeseries, so may even pick it up (unless my other approach works out nicely...)

**Pull request closed without merging**